### DISCUSSION TITLE
Conforming Delaunay/Voronoi with Z Heights

### DISCUSSION THREAD

**Discussion by bVess12** [NetTopologySuite/NetTopologySuite#729] [Q&A]
Hi, I'm new to NTS and am having trouble finding information.  I'm looking to (quickly) get a list of delaunay triangles/voronoi points within a set of polygons/holes AND get the following information:

1. Get a list of voronoi sites/delaunay circumcenters and their neighbors that are INSIDE the given constraints (like this image)
![image](https://github.com/NetTopologySuite/NetTopologySuite/assets/43218369/5b811a5f-5912-4f46-80d6-ba1db880408a)

2. For each voronoi site/delaunay circumcenter I need to know the shortest distance to the polygon/holes that made up the constraints

I thought I had all of this information by using the CoordinateM (I expected the M coordinate to be distance from the outer polygon), but I must be misunderstanding something, because the M/Z values of the resulting triangle are always either 0 or NaN.  I've attached my stab at this below. Please let me know if there's something I'm missing here!

```C#
NetTopologySuite.Geometries.Polygon inputPolygon;

var fac = NtsGeometryServices.Instance.CreateGeometryFactory();
LinearRing toLinearRing(Polyline3D p)
{
    if (!p.IsClosed)
        throw new Exception("Can't convert open polyline to linear ring");
    return 
        LinearRing.DefaultFactory.CreateLinearRing(
            p.GetAllPoints().Select(v => new CoordinateM(v.X, v.Y, 0)).ToArray());
}

//Generate the input geometry
if (polylineWithHoles.Holes != null && polylineWithHoles.Holes.Any())
{
    inputPolygon =
        NetTopologySuite.Geometries.Polygon.DefaultFactory.CreatePolygon(
            toLinearRing(polylineWithHoles.Polyline),
            polylineWithHoles.Holes.Select(h => toLinearRing(h)).ToArray());
}
else
    inputPolygon =
        NetTopologySuite.Geometries.Polygon.DefaultFactory.CreatePolygon(
            toLinearRing(polylineWithHoles.Polyline));

//Generate a delaunay triangulation
var builder = new ConformingDelaunayTriangulationBuilder();
builder.SetSites(inputPolygon);
builder.Constraints = inputPolygon;
var result = builder.GetTriangles(GeometryCollection.DefaultFactory);
foreach(var g in result)
{
    var polygon = (NetTopologySuite.Geometries.Polygon)g;
             
    foreach(var coord in polygon.Coordinates)
    {
        string msg = $"({coord.X}, {coord.Y}, {coord.Z}, {coord.M}";
        if (!double.IsNaN(coord.Z) || !double.IsNaN(coord.M))
            if(coord.M != 0)
                Console.WriteLine(msg);
    }
}
```

**Comment by FObermaier** [comment]
That is a misunderstanding on your part. The measure value on `CoordinateM` or `CoordinateZM` is a value that you have to provide, nothing that is computed within NetTopologySuite. It can be just about anything, not necessarily a distance value.
